### PR TITLE
ICSAAS-303 implement typeahead service

### PR DESCRIPTION
**Pull request by igorsimovski**
- fix javadocs typos
 - define public interfaces for typeahead feature
 - add javadocs
 - fix typo in DefaultFulltextSearchService osgi configs property
 - make existing extraction data strategy implementations final
 - implement extraction data strategy for typeahead data
 - implement typeahead text parser
  - parse/resolve the term and prefix from the search text
 - implement a TypeaheadTextQuery
 - define a typeahead payload builder
 - add unit tests

### CODE REVIEW

**Comment by igorsimovski** [comment]
warning: PR is branched out from https://github.com/valtech-ch/saas-aem-module/pull/17
ideally we first merge 306, and then this one.

File: api/src/main/java/com/valtech/aem/saas/api/typeahead/TypeaheadPayload.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
+package com.valtech.aem.saas.api.typeahead;
+
+import java.util.Map;
+
+public interface TypeaheadPayload {
```
Comment (ejanevska) [inline api/src/main/java/com/valtech/aem/saas/api/typeahead/TypeaheadPayload.java:5]: javadoc missing

File: core/src/main/java/com/valtech/aem/saas/core/query/TypeaheadSearchTextParser.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
+package com.valtech.aem.saas.core.query;
+
+import java.util.Optional;
+import lombok.RequiredArgsConstructor;
+import org.apache.commons.lang3.StringUtils;
+
+@RequiredArgsConstructor
```
Comment (ejanevska) [inline core/src/main/java/com/valtech/aem/saas/core/query/TypeaheadSearchTextParser.java:7]: javadoc is missing

File: core/src/main/java/com/valtech/aem/saas/core/query/TypeaheadTextQuery.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
+package com.valtech.aem.saas.core.query;
+
+import com.valtech.aem.saas.api.query.TypeaheadQuery;
+import java.util.ArrayList;
+import java.util.List;
+import org.apache.commons.collections4.CollectionUtils;
+import org.apache.http.NameValuePair;
+import org.apache.http.message.BasicNameValuePair;
+
+public class TypeaheadTextQuery implements TypeaheadQuery {
```
Comment (ejanevska) [inline core/src/main/java/com/valtech/aem/saas/core/query/TypeaheadTextQuery.java:10]: javadoc is missing

File: core/src/main/java/com/valtech/aem/saas/core/typeahead/IndexTypeaheadConsumerService.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+import java.util.List;
+import java.util.stream.Collectors;
+import lombok.Builder;
+import lombok.NonNull;
+import lombok.Singular;
+import org.apache.commons.collections4.CollectionUtils;
+import org.apache.commons.collections4.MapUtils;
+import org.apache.commons.lang3.StringUtils;
+
+@Builder
```
Comment (ejanevska) [inline core/src/main/java/com/valtech/aem/saas/core/typeahead/IndexTypeaheadConsumerService.java:23]: javadoc missing

File: core/src/test/java/com/valtech/aem/saas/core/query/TypeaheadSearchTextParserTest.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    MatcherAssert.assertThat(properParser.getTerm().get(), is("bar"));
+
+    TypeaheadSearchTextParser properParserWithTrailingWhiteSpaces = new TypeaheadSearchTextParser("foo  bar  ");
+    MatcherAssert.assertThat(properParserWithTrailingWhiteSpaces.getPrefix().isPresent(), is(true));
+    MatcherAssert.assertThat(properParserWithTrailingWhiteSpaces.getPrefix().get(), is("foo"));
+    MatcherAssert.assertThat(properParserWithTrailingWhiteSpaces.getTerm().isPresent(), is(true));
+    MatcherAssert.assertThat(properParserWithTrailingWhiteSpaces.getTerm().get(), is("bar"));
+
+    TypeaheadSearchTextParser noPrefixParser = new TypeaheadSearchTextParser("foo");
+    MatcherAssert.assertThat(noPrefixParser.getPrefix().isPresent(), is(false));
```
Comment (ejanevska) [inline core/src/test/java/com/valtech/aem/saas/core/query/TypeaheadSearchTextParserTest.java:25]: we can use static import in all tests

File: core/src/main/java/com/valtech/aem/saas/core/typeahead/IndexTypeaheadConsumerService.java (hunk 4 above)
Comment (igorsimovski) [reply core/src/main/java/com/valtech/aem/saas/core/typeahead/IndexTypeaheadConsumerService.java:23]: not needed

File: core/src/main/java/com/valtech/aem/saas/core/query/TypeaheadSearchTextParser.java (hunk 2 above)
Comment (igorsimovski) [reply core/src/main/java/com/valtech/aem/saas/core/query/TypeaheadSearchTextParser.java]: not needed

**Review by ejanevska — APPROVED** [review]

File: core/src/main/java/com/valtech/aem/saas/core/typeahead/DefaultTypeaheadService.java [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+    @AttributeDefinition(name = "Api action",
+        description = "Path designating the action",
+        type = AttributeType.STRING)
+    String typeaheadService_apiAction() default DEFAULT_API_ACTION;
+
+    @AttributeDefinition(name = "Allowed filter fields",
+        description = "List of field names that can be used in filter queries",
+        type = AttributeType.STRING)
+    String[] typeaheadService_allowedFilterFields() default {
+        "language",
```
Comment (nhirrle) [inline core/src/main/java/com/valtech/aem/saas/core/typeahead/DefaultTypeaheadService.java:94]: - [ ] Please move those values to individual constants
- [ ] Please keep only the non custom filters in here: repository_path_url, language, domain. The custom ones such as page_type_str, categories_murl,... are consumer configs

File: core/src/main/java/com/valtech/aem/saas/core/http/response/TypeaheadDataExtractionStrategy.java [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    return Optional.ofNullable(facetCounts.getFacetFields())
+        .filter(MapUtils::isNotEmpty)
+        .map(map -> map.get(getTypeaheadSolrIndexFieldName()))
+        .map(this::getTypeaheadItems)
+        .orElse(Collections.emptyList());
+  }
+
+  private List<String> getTypeaheadItems(@NonNull List<Object> facetFieldsList) {
+    List<String> items = new ArrayList<>();
+    for (int i = 0; i < facetFieldsList.size(); i++) {
```
Comment (nhirrle) [inline core/src/main/java/com/valtech/aem/saas/core/http/response/TypeaheadDataExtractionStrategy.java:48]: this looks weired. can you provide an example? I don't get why odd elements behave differently

File: core/src/main/java/com/valtech/aem/saas/core/query/TypeaheadTextQuery.java [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+  public static final String REGEX_MATCHING_CONSECUTIVE_WHITESPACE_CHARS = "\\s{2,}";
+
+  public static final String KEY_TERM = "term";
+  public static final String KEY_PREFIX = "prefix";
+
+  private final NameValuePair term;
+  private NameValuePair prefix;
+
+  public TypeaheadTextQuery(String searchText) {
+    term = new BasicNameValuePair(KEY_TERM, "*");
```
Comment (nhirrle) [inline core/src/main/java/com/valtech/aem/saas/core/query/TypeaheadTextQuery.java:27]: please use the constant for "*" . I think you defined it already once

File: core/src/main/java/com/valtech/aem/saas/core/typeahead/IndexTypeaheadConsumerService.java [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+public final class IndexTypeaheadConsumerService implements TypeaheadConsumerService {
```
Comment (nhirrle) [inline core/src/main/java/com/valtech/aem/saas/core/typeahead/IndexTypeaheadConsumerService.java:24]: same as in https://github.com/valtech-ch/saas-aem-module/pull/17#discussion_r727181927 - please migrate to an osgi service

File: core/src/main/java/com/valtech/aem/saas/core/http/response/TypeaheadDataExtractionStrategy.java (hunk 7 above)
Comment (igorsimovski) [reply core/src/main/java/com/valtech/aem/saas/core/http/response/TypeaheadDataExtractionStrategy.java:48]: check file -> core/src/test/resources/__files/search/typeahead/success.json for example

File: core/src/main/java/com/valtech/aem/saas/core/common/resource/ParentResource.java [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+import lombok.RequiredArgsConstructor;
+import org.apache.sling.api.resource.Resource;
+
+@RequiredArgsConstructor
+public final class ParentResource {
+
+  @NonNull
+  private final Resource resource;
+
+  public Optional<Resource> getParentWithResourceType(String resourceType) {
```
Comment (nhirrle) [inline core/src/main/java/com/valtech/aem/saas/core/common/resource/ParentResource.java:14]: for what do we need this method?  I only see tests for it but no usage

File: core/src/main/java/com/valtech/aem/saas/core/common/resource/ResourceChildren.java [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+import lombok.RequiredArgsConstructor;
+import org.apache.sling.api.resource.Resource;
+
+@RequiredArgsConstructor
+public final class ResourceChildren {
+
+  @NonNull
+  private final Resource resource;
+
+  public Stream<Resource> getDirectChildren() {
```
Comment (nhirrle) [inline core/src/main/java/com/valtech/aem/saas/core/common/resource/ResourceChildren.java:15]: for what do we need this method? I only see tests for it but no usage

File: core/src/main/java/com/valtech/aem/saas/core/fulltextsearch/SearchResultsServlet.java [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+import org.apache.sling.api.resource.ResourceResolver;
+import org.apache.sling.api.servlets.HttpConstants;
+import org.apache.sling.api.servlets.SlingSafeMethodsServlet;
+import org.apache.sling.servlets.annotations.SlingServletResourceTypes;
+import org.osgi.service.component.annotations.Component;
+
+@Slf4j
+@Component(service = Servlet.class)
+@SlingServletResourceTypes(
+    resourceTypes = SearchResultsServlet.RESOURCE_TYPE_CQ_PAGE,
```
Comment (nhirrle) [inline core/src/main/java/com/valtech/aem/saas/core/fulltextsearch/SearchResultsServlet.java:31]: the request should not be bound to the page but it should be bound to the search component. you could have as well several search components on the page (in theory)

File: core/src/main/java/com/valtech/aem/saas/core/http/response/TypeaheadDataExtractionStrategy.java (hunk 7 above)
Comment (nhirrle) [reply core/src/main/java/com/valtech/aem/saas/core/http/response/TypeaheadDataExtractionStrategy.java:48]: thank you! weired solr response :) 
can you extract the if condition into a method `isTypeaheadTextOption(facetFieldsList.get(i))` so it get's more readable?

File: core/src/main/java/com/valtech/aem/saas/core/http/response/TypeaheadDataExtractionStrategy.java [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+import lombok.RequiredArgsConstructor;
+import org.apache.commons.collections4.MapUtils;
+
+/**
+ * A strategy for extracting typeahead options.
+ */
+@RequiredArgsConstructor
+public final class TypeaheadDataExtractionStrategy implements SearchResponseDataExtractionStrategy<List<String>> {
+
+  private static final String TYPEAHEAD_SOLR_INDEX_FIELD_NAME_PREFIX = "autocomplete_text_";
```
Comment (nhirrle) [inline core/src/main/java/com/valtech/aem/saas/core/http/response/TypeaheadDataExtractionStrategy.java:19]: please remove `solr` out of the constant and getTypeaheadSolrIndexFieldName . We don't need to expose the technology

File: core/src/main/java/com/valtech/aem/saas/core/i18n/I18nProvider.java [hunk 14]
```diff
@@ -0,0 +1,10 @@
+package com.valtech.aem.saas.core.i18n;
+
+import com.day.cq.i18n.I18n;
+import lombok.NonNull;
+import org.apache.sling.api.SlingHttpServletRequest;
+
+public interface I18nProvider {
```
Comment (nhirrle) [inline core/src/main/java/com/valtech/aem/saas/core/i18n/I18nProvider.java:7]: javadoc please

File: core/src/main/java/com/valtech/aem/saas/core/i18n/I18nProvider.java [hunk 15]
```diff
@@ -0,0 +1,10 @@
...
+
+  I18n getI18n(@NonNull SlingHttpServletRequest request);
```
Comment (nhirrle) [inline core/src/main/java/com/valtech/aem/saas/core/i18n/I18nProvider.java:9]: javadoc

File: core/src/main/java/com/valtech/aem/saas/core/typeahead/DefaultTypeaheadService.java [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+
+    @AttributeDefinition(name = "Api action",
+        description = "Path designating the action",
+        type = AttributeType.STRING)
+    String typeaheadService_apiAction() default DEFAULT_API_ACTION;
+
+    @AttributeDefinition(name = "Allowed filter fields",
+        description = "List of field names that can be used in filter queries",
+        type = AttributeType.STRING)
+    String[] typeaheadService_allowedFilterFields() default {
```
Comment (nhirrle) [inline core/src/main/java/com/valtech/aem/saas/core/typeahead/DefaultTypeaheadService.java:140]: I think we discussed this once: Is it required limit those filter really? Worst case search will respond with an error. Reason I am asking: Fields can be added in runtime on search admin. 
We don't want to have a new aem release than as ogi configs are immutable so we would need through cloud manager

File: ui.apps/src/main/content/jcr_root/apps/saas-aem-module/i18n/en/com.valtech.aem.saas.core.search.loadmore.button.label/.content.xml [hunk 17]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<jcr:root
+    xmlns:jcr="http://www.jcp.org/jcr/1.0"
+    xmlns:sling="http://sling.apache.org/jcr/sling/1.0"
+    jcr:mixinTypes="[sling:Message]"
+    jcr:primaryType="sling:MessageEntry"
+    sling:message="Load more"/>
```
Comment (nhirrle) [inline ui.apps/src/main/content/jcr_root/apps/saas-aem-module/i18n/en/com.valtech.aem.saas.core.search.loadmore.button.label/.content.xml:7]: please use the new i18n json format instead

**Pull request merged**